spynnaker/pyNN/extra_algorithms/neuron_expander.py

```diff
@@ -0,0 +1,107 @@
+        Needs to be done after data has been loaded.
+    """
+    neuron_bin = SpynnakerDataView.get_executable_path(NEURON_EXPANDER_APLX)
+
+    # Find the places where the neuron expander should run
+    expander_cores, expanded_pop_vertices = _plan_expansion(neuron_bin)
+
+    progress = ProgressBar(expander_cores.total_processors,
+                           "Expanding Neuron Data")
+    expander_app_id = SpynnakerDataView.get_new_id()
+    run_system_application(
+        expander_cores, expander_app_id,
+        get_config_bool("Reports", "write_expander_iobuf"),
+        None, [CPUState.FINISHED], False,
+        "neuron_expander_on_{}_{}_{}.txt", progress_bar=progress,
+        logger=logger)
+    progress.end()
+
+    _fill_in_initial_data(expanded_pop_vertices)
+
+
+def _plan_expansion(synapse_expander_bin):
+    """ Plan the expansion of synapses and set up the regions using USER1
+
+    :param ~pacman.model.placements.Placements: The placements of the vertices
+    :param str synapse_expander_bin: The binary name of the synapse expander
+    :param ~spinnman.transceiver.Transceiver transceiver:
+        How to talk to the machine
+    :return: The places to load the synapse expander and delay expander
+        executables, and the target machine vertices to read synapses back from
+    :rtype: (ExecutableTargets, list(MachineVertex, Placement))
+    """
+    expander_cores = ExecutableTargets()
+    expanded_pop_vertices = list()
+
+    progress = ProgressBar(
+        SpynnakerDataView.get_n_placements(),
+        "Preparing to Expand Neuron Data")
+    for placement in progress.over(SpynnakerDataView.iterate_placemements()):
+        # Add all machine vertices of the population vertex to ones
+        # that need synapse expansion
+        vertex = placement.vertex
+        if isinstance(vertex, AbstractNeuronExpandable):
+            if vertex.gen_neurons_on_machine():
+                expanded_pop_vertices.append((vertex, placement))
+                expander_cores.add_processor(
+                    synapse_expander_bin,
+                    placement.x, placement.y, placement.p,
+                    executable_type=ExecutableType.SYSTEM)
+                # Write the region to USER1, as that is the best we can do
+                write_address_to_user1(
+                    placement.x, placement.y, placement.p,
+                    vertex.neuron_generator_region)
+
+    return expander_cores, expanded_pop_vertices
+
+
+def _fill_in_initial_data(expanded_pop_vertices):
+    """ Once expander has run, fill in the connection data
+
+    :param list(MachineVertex, Placement) expanded_pop_vertices:
+        List of machine vertices to read data from
+    :param ~spinnman.transceiver.Transceiver transceiver:
+        How to talk to the machine
+
+    :rtype: None
+    """
+    progress = ProgressBar(
+        len(expanded_pop_vertices), "Getting initial values")
+    for vertex, placement in progress.over(expanded_pop_vertices):
```

Comment:
Do you really need three ProgressBars here?

Would one bar with three steps is not be enough here

Reply:
This is much more likely to be slow, as it goes vertex-by-vertex